src/Core/Domain/Country/QueryResult/Countries.php

```diff
@@ -0,0 +1,54 @@
+<?php
+/**
+ * 2007-2019 PrestaShop and Contributors
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.txt.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2019 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+namespace PrestaShop\PrestaShop\Core\Domain\Country\QueryResult;
+
+/**
+ * Holds list of countries found by query configuration
+ */
+class Countries
```

Comment:
I feel like this is too generic, @matks wdyt?

Reply:
Mmmm 🤔 I'm afraid it's just ... countries. Nothing more. Cant think of something better 😅

Reply:
Actually there is a huge difference. Image you have 2 different use cases:
1. For X case, were we need to get countries with id, name, state and addresses that belongs to it.
2. For Y case, need to get countries with id, states, customers that belongs to it.

As a result we have 2 different queries:

```php
interface GetCountriesForXHandlerInterface {
    public function handle(GetCountriesForX $query): CountriesForX
}

interface GetCountriesForYHandlerInterface {
    public function handle(GetCountriesForY $query): CountriesForY
}
```

We even have examples in current implementation, both `GetCustomerForEditing` and `GetCustomerForViewing` returns the same customer, but with totally different data.

This is what builds PrestaShop's API, not a generic `GetCountries` query which returns everything, but a specialized rich domain model queries.

At least this is how I see it. :stuck_out_tongue:

Reply:
I think we should then define
- a Country VO with Country dedicated data
- another object (we need to find a suitable name) to describe the data needed for X, that will probably be a data structure that contains some country VOs
- another object (we need to find a suitable name) to describe the data needed for Y, that will probably be a data structure that contains some country VOs

Reply:
Not really, I think it's enough to have `Country/QueryResult/CountryForX` and `Country/QueryResult/CountryForY`, it's just a DTO, we don't have any business rules to protect on read side.

Regarding `Country` VO, I don't think it makes sense at this point, in the end, it's not a VO, bu entity and as far as I understand entities will be introduced in future major version (I guess v100.0.0 or later :man_shrugging: ). :)

Reply:
In the current implementation - only CountryByIdConfigurableChoiceProvider is using GetCountries Query. It requires only country id and name. So created Country VO with id and name. Also ArrayCollection of these VO's.